android/src/main/java/com/pagopa/ioreactnativecrypto/X509VerificationUtil.kt

```diff
@@ -0,0 +1,428 @@
+        errorMessage = baseErrorMessage,
+        failingCertificate = failingCert
+      )
+    }
+  }
+
+  /**
+   * Handles CertPathValidatorException on API Level 24+ using BasicReason.
+   * This function is marked with @RequiresApi(24) and should only be called
+   * when Build.VERSION.SDK_INT >= Build.VERSION_CODES.N.
+   */
+  @RequiresApi(Build.VERSION_CODES.N)
+  private fun handleCertPathValidatorExceptionApi24(
+    e: CertPathValidatorException,
+    failingCert: X509Certificate?,
+    baseErrorMessage: String
+  ): ValidationResult {
+    val reason = e.reason
+
+    val status = when (reason) {
+      CertPathValidatorException.BasicReason.REVOKED -> ValidationStatus.CERTIFICATE_REVOKED
+      CertPathValidatorException.BasicReason.UNDETERMINED_REVOCATION_STATUS -> ValidationStatus.CRL_FETCH_FAILED // Indicates revocation check couldn't complete
+      CertPathValidatorException.BasicReason.EXPIRED -> ValidationStatus.CERTIFICATE_EXPIRED // Should have been caught earlier, but handle defensively
+      CertPathValidatorException.BasicReason.NOT_YET_VALID -> ValidationStatus.CERTIFICATE_NOT_YET_VALID // Should have been caught earlier
+      CertPathValidatorException.BasicReason.INVALID_SIGNATURE -> ValidationStatus.INVALID_CHAIN_PATH
+      CertPathValidatorException.BasicReason.ALGORITHM_CONSTRAINED -> ValidationStatus.INVALID_CHAIN_PATH // Algorithm policy issue
+      CertPathValidatorException.BasicReason.UNSPECIFIED -> ValidationStatus.INVALID_CHAIN_PATH // General path issue
+      else -> ValidationStatus.INVALID_CHAIN_PATH // Default for any other/new reasons
+    }
+
+    val detailedErrorMessage = "$baseErrorMessage (Reason: $reason)"
+    return ValidationResult(false, status, detailedErrorMessage, failingCert)
+  }
+
+
+  /** Internal exception class for CRL fetching issues. */
+  private class CrlFetchException(val status: ValidationStatus, message: String, cause: Throwable? = null) : IOException(message, cause)
+
+  /**
+   * Fetches CRLs specified in the CRL Distribution Points extension of certificates in the chain.
+   * It tries multiple URLs if available and validates the CRL's validity period and signature.
+   * This is only called if at least one certificate in the chain has a CDP.
+   */
+  private suspend fun fetchCrlsForChain(
+    factory: CertificateFactory,
+    chain: List<X509Certificate>,
+    trustAnchorCert: X509Certificate, // Needed for CRL signature verification
+    options: X509VerificationOptions
+  ): List<X509CRL> {
+
+    val uniqueCrlUrls = chain.filter { hasCrlDistributionPoint(it) }
+      .flatMap { extractCrlDistributionPoints(it) }
+      .distinct()
+
+    if (uniqueCrlUrls.isEmpty()) {
+      return emptyList()
+    }
+
+    val validCrls = mutableListOf<X509CRL>()
+    var lastException: CrlFetchException? = null // Track the last significant error
+
+    for (url in uniqueCrlUrls) {
+      try {
+        val crlBytes = downloadCrlWithTimeout(url, options)
```

Comment:
Instead of downloading the CRL every time, the best practice is to download it once and save it for future checks. And add a mechanism to re-download it only after a configurable period of time. This is to avoid excessive downloads in a short period of time considering that the CRLs could be very large.